[Feat]: launch.json for debugging appium and its unit test from source

I am trying to understand what's going on in the Appium source code. And it's really tough without debugging tool.

Maybe we can add one more configuration to `launch.json` for both debugging appium and its test

- So far I got one for debug appium.

```
{
      "type": "node",
      "request": "launch",
      "name": "launch appium",
      "skipFiles": ["<node_internals>/**"],
      "cwd": "${workspaceFolder}",
      "program": "${workspaceFolder}/packages/appium/index.js",
      "internalConsoleOptions": "openOnSessionStart",
    },
```
this one is working

- For the unit test, I tried this config

```
{
      "type": "node",
      "request": "launch",
      "name": "Mocha Tests",
      "program": "${workspaceFolder}/node_modules/mocha/bin/_mocha",
      "args": [
        "--no-timeouts",
        "--require",
        "ts-node/register",
        "${workspaceFolder}/packages/appium/test/unit/extension/*.spec.js"
      ],
    }
```

It doesn't work yet.

@mykola-mokhnach `Test - Current File` in https://github.com/appium/appium/blob/master/.vscode/launch.json looks like what I am looking for.

However, when I run it I got `Uncaught Error Error: Cannot find module '@babel/register'
Require stack:`. Maybe there is an additional step to do.

The reason I ask is that the latest unit test is failing on my machine, more info here https://discuss.appium.io/t/the-unit-test-failed-at-8db0122/37971

--- a/.vscode/launch.json
+++ b/.vscode/launch.json
@@ -1,32 +1,62 @@
 {
-  "version": "0.2.0",
   "configurations": [
     {
-      "args": ["--colors", "--timeout", "0", "${file}"],
-      "internalConsoleOptions": "openOnSessionStart",
-      "name": "Test - Current File",
-      "program": "${workspaceFolder}/node_modules/mocha/bin/_mocha",
-      "runtimeArgs": ["--require", "@babel/register"],
-      "request": "launch",
-      "skipFiles": ["<node_internals>/**", "**/node_modules/mocha/**"],
-      "type": "node",
-      "sourceMaps": true,
+      "args": [
+        "--log-level=debug"
+      ],
+      "console": "integratedTerminal",
       "cwd": "${workspaceFolder}",
       "env": {
-        "NODE_ENV": "test",
-        "BABEL_ENV": "test"
+        "_FORCE_LOGS": "1"
       },
-      "autoAttachChildProcesses": true,
+      "internalConsoleOptions": "openOnSessionStart",
+      "name": "Run Appium",
+      "program": "${workspaceFolder}/node_modules/.bin/appium",
+      "request": "launch",
+      "skipFiles": [
+        "<node_internals>/**"
+      ],
+      "sourceMaps": true,
+      "type": "node"
     },
     {
+      "args": [
+        "--colors",
+        "--no-timeout",
+        "${file}"
+      ],
+      "autoAttachChildProcesses": true,
+      "console": "integratedTerminal",
+      "cwd": "${workspaceFolder}",
       "internalConsoleOptions": "openOnSessionStart",
-      "name": "TypeDoc",
+      "name": "Test Current File w/ Mocha",
+      "program": "${workspaceFolder}/node_modules/.bin/mocha",
+      "request": "launch",
+      "runtimeArgs": [
+        "--require",
+        "ts-node/register"
+      ],
+      "skipFiles": [
+        "<node_internals>/**",
+        "**/node_modules/mocha/**",
+        "**/node_modules/ts-node/**"
+      ],
+      "sourceMaps": true,
+      "type": "node"
+    },
+    {
+      "console": "integratedTerminal",
+      "cwd": "${workspaceFolder}",
+      "internalConsoleOptions": "openOnSessionStart",
+      "name": "Typedoc",
       "program": "${workspaceFolder}/node_modules/.bin/typedoc",
       "request": "launch",
-      "skipFiles": ["<node_internals>/**"],
-      "type": "node",
+      "skipFiles": [
+        "<node_internals>/**"
+      ],
       "sourceMaps": true,
-      "cwd": "${workspaceFolder}",
+      "type": "node"
     }
-  ]
+  ],
+  "version": "0.2.0"
 }

--- a/packages/gulp-plugins/.mocharc.js
+++ b/packages/gulp-plugins/.mocharc.js
@@ -1,0 +1,11 @@
+// @ts-check
+
+'use strict';
+
+module.exports = {
+  require: [require.resolve('./test/setup.js')],
+  // forbids use of .only() in CI
+  forbidOnly: Boolean(process.env.CI),
+  // increase default timeout for CI since it can be slow
+  timeout: process.env.CI ? '5s' : '2s',
+};

--- a/packages/gulp-plugins/package.json
+++ b/packages/gulp-plugins/package.json
@@ -35,9 +35,9 @@
     "build"
   ],
   "scripts": {
+    "build": "gulp transpile",
     "clean": "npx rimraf build",
     "dev": "gulp dev --no-notif",
-    "build": "gulp transpile",
     "prepare": "npm run build",
     "test": "gulp unit-test:run",
     "test:e2e": "gulp e2e-test:run",
@@ -93,6 +93,9 @@
     "through2": "4.0.2",
     "vinyl-paths": "3.0.1",
     "yargs": "17.6.2"
+  },
+  "devDependencies": {
+    "@babel/register": "7.18.9"
   },
   "peerDependencies": {
     "gulp": "^4.0.2"

--- a/packages/gulp-plugins/test/setup.js
+++ b/packages/gulp-plugins/test/setup.js
@@ -1,0 +1,32 @@
+'use strict';
+
+/**
+ * Mocha will load this file to configure the environment to use Chai as the
+ * assertion library. Since Chai is a singleton, we can run into problems when
+ * running files individually, if we have not carefully configured Chai in every
+ * single test file. This file means less boilerplate and less random test
+ * failures when running single test files.
+ *
+ * For simplicity, this file is _not_ transpiled.  If it were, Mocha would need
+ * to load different versions of this file depending on the test context (are we
+ * running tests against the distfiles, or the source files?).
+ *
+ */
+
+// This configures @babel/register to look for a babel config in parent dir(s) of
+// wherever the tests are run.  This is required if running tests via `mocha` in a package dir
+// instead of the monorepo root. This does not affect `gulp-mocha`, since it has its own `.babelrc`.
+// This file is required _in addition to_ the monorepo root's `test/setup.js`.
+
+require('@babel/register')();
+
+const chai = require('chai');
+const chaiAsPromised = require('chai-as-promised');
+const sinonChai = require('sinon-chai');
+
+// The `chai` global is set if a test needs something special.
+// Most tests won't need this.
+global.chai = chai.use(chaiAsPromised).use(sinonChai);
+
+// `should()` is only necessary when working with some `null` or `undefined` values.
+global.should = chai.should();

--- a/packages/gulp-plugins/test/transpile-specs.js
+++ b/packages/gulp-plugins/test/transpile-specs.js
@@ -90,7 +90,7 @@ describe('transpile-specs', function () {
 
       it(`should use sourcemap when throwing within mocha (${name})`, async function () {
         const [stdout, stderr] = await exec(
-          `${MOCHA} build-fixtures/test/${files.classFile}-throw-specs.js`
+          `${MOCHA} --no-color build-fixtures/test/${files.classFile}-throw-specs.js`
         );
         print(stdout, stderr);
         let output = stdout + stderr;

--- a/tsconfig.json
+++ b/tsconfig.json
@@ -1,12 +1,15 @@
 {
   "$schema": "https://json.schemastore.org/tsconfig",
   "extends": "@tsconfig/node14/tsconfig.json",
+  "files": [],
   "ts-node": {
     "transpileOnly": true
   },
-  "files": [],
   "compilerOptions": {
-    "noErrorTruncation": true
+    "allowJs": true,
+    "allowSyntheticDefaultImports": true,
+    "resolveJsonModule": true,
+    "noEmit": true
   },
   "references": [
     {

